amm-executor/src/Spectrum/Executor/Backlog/Service.hs

```diff
@@ -1,60 +1,158 @@
+    , suspend = \(SuspendedOrder order timestamp) -> do
+        modifyIORef suspendedPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+        isExist $ BacklogOrder timestamp order
+    , checkLater = \(InProgressOrder order timestamp) -> do
+        modifyIORef toRevisitQ (mkWeightedOrderWithTimestamp order timestamp Seq.<|)
+        isExist $ BacklogOrder timestamp order
+    , tryAcquire = do
+        _ <- refreshQueues config store pendingPQ suspendedPQ toRevisitQ
+        maxWeightedOrderM <- getMaxWeightedOrder' pendingPQ suspendedPQ
+        case maxWeightedOrderM of
+          Just (WeightedOrderWithTimestamp oId _ _) -> get oId <&> fmap backlogOrder
+          Nothing -> pure Nothing
+    , drop = \(ExecutedOrder oId) -> do
+        _ <- modifyIORef pendingPQ   (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef suspendedPQ (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef toRevisitQ  (Seq.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        drop oId
+    }
 
-  (_, db) <- Rocks.openBracket storePath
-              Rocks.defaultOptions
-                { Rocks.createIfMissing = createIfMissing
-                }
+refreshQueues
+  :: (MonadIO m)
+  => BacklogServiceConfig
+  -> BacklogStore m
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (Seq.Seq WeightedOrderWithTimestamp)
+  -> m ()
+refreshQueues BacklogServiceConfig{..} BacklogStore{..} pendingQueueRef suspendedQueueRef toRevisitSeqRef = do
+  currentTime <- getCurrentTime
+  revisited2process <-
+    atomicModifyIORef'
+      toRevisitSeqRef
+      (Seq.partition (\(WeightedOrderWithTimestamp _ _ oTime) -> diffUTCTime currentTime oTime < orderExecTime))
+  suspended2drop   <-
+    atomicModifyIORef'
+      suspendedQueueRef
+      (PQ.partition (\(WeightedOrderWithTimestamp _ _ oTime) -> diffUTCTime currentTime oTime < orderLifetime))
+  _ <- (\order@(WeightedOrderWithTimestamp oId _ oTime) ->
+          if diffUTCTime currentTime oTime < orderLifetime
+            then modifyIORef' pendingQueueRef (PQ.insert order)
+            else drop oId
+       ) `traverse` revisited2process
+  _ <- (\(WeightedOrderWithTimestamp oId _ _) -> drop oId) `traverse` PQ.toList suspended2drop
+  pure ()
+
+getMaxWeightedOrder'
+  :: (MonadIO m)
+  => IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> m (Maybe WeightedOrderWithTimestamp)
+getMaxWeightedOrder' pendingQueueRef suspendedQueueRef = do
+  pendingQueue   <- readIORef pendingQueueRef
+  suspendedQueue <- readIORef suspendedQueueRef
+  let
+    maxPendingM   = PQ.getMax pendingQueue
+    maxSuspendedM = PQ.getMax suspendedQueue
+  case (maxPendingM, maxSuspendedM) of
+    (Just maxPending, Just maxSuspended) ->
+      if maxPending > maxSuspended
```

Comment:
Suspended orders should be picked with some configured probability. Why this condition?

Reply:
What do you mean under "probability" for orders? We don't have any additional information about pool state at this moment. On my opinion the best way to select the most preferredable order is weight compare

Reply:
So you should pick orders for retry in N of M cases

Reply:
no, all orders have timestamp (time when they have been parsed from ledger). When diff between (currentTime - orderTime) > orderLifetime, order will be discarded

Reply:
Bc other orders will wait until then

Reply:
Each cycle the same order with the highest weight will be picked until it's expired